wrong "fruitless type test" warning depending on source file order

Scala 2.13.10

```
➜ sandbox cat A.scala
object Foo {
  sealed trait A
  sealed trait B
  final case class C(x: Int) extends A with B
}

➜ sandbox cat B.scala
object T {
  import Foo._
  def f(a: A) = a match {
    case b: B => 0
    case _ => 1
  }
}

➜ sandbox sc A.scala B.scala

➜ sandbox sc B.scala A.scala
B.scala:4: warning: fruitless type test: a value of type Foo.A cannot also be a Foo.B
    case b: B => 0
            ^
1 warning
```

--- a/src/compiler/scala/tools/nsc/CompilationUnits.scala
+++ b/src/compiler/scala/tools/nsc/CompilationUnits.scala
@@ -127,7 +127,9 @@ trait CompilationUnits { global: Global =>
     val transformed = new mutable.AnyRefMap[Tree, Tree]
 
     /** things to check at end of compilation unit */
-    val toCheck = new ListBuffer[() => Unit]
+    val toCheck = new ListBuffer[CompilationUnit.ToCheck]
+    private[nsc] def addPostUnitCheck(check: CompilationUnit.ToCheckAfterUnit): Unit = toCheck += check
+    private[nsc] def addPostTyperCheck(check: CompilationUnit.ToCheckAfterTyper): Unit = toCheck += check
 
     /** The features that were already checked for this unit */
     var checkedFeatures = Set[Symbol]()
@@ -148,5 +150,11 @@ trait CompilationUnits { global: Global =>
     val isJava: Boolean = source.isJava
 
     override def toString() = source.toString()
+  }
+
+  object CompilationUnit {
+    sealed trait ToCheck
+    trait ToCheckAfterUnit extends ToCheck { def apply(): Unit }
+    trait ToCheckAfterTyper extends ToCheck { def apply(): Unit }
   }
 }

--- a/src/compiler/scala/tools/nsc/typechecker/Analyzer.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/Analyzer.scala
@@ -13,6 +13,8 @@
 package scala.tools.nsc
 package typechecker
 
+import scala.collection.mutable.ListBuffer
+
 /** Defines the sub-components for the namer, packageobjects, and typer phases.
  */
 trait Analyzer extends AnyRef
@@ -98,15 +100,19 @@ trait Analyzer extends AnyRef
       // Lacking a better fix, we clear it here (before the phase is created, meaning for each
       // compiler run). This is good enough for the resident compiler, which was the most affected.
       undoLog.clear()
+      private val toCheckAfterTyper = new ListBuffer[CompilationUnit.ToCheckAfterTyper]
       override def run(): Unit = {
         val start = if (settings.areStatisticsEnabled) statistics.startTimer(statistics.typerNanos) else null
         global.echoPhaseSummary(this)
         val units = currentRun.units
+
         while (units.hasNext) {
           applyPhase(units.next())
           undoLog.clear()
         }
         finishComputeParamAlias()
+        for (work <- toCheckAfterTyper) work()
+        toCheckAfterTyper.clear()
         // defensive measure in case the bookkeeping in deferred macro expansion is buggy
         clearDelayed()
         if (settings.areStatisticsEnabled) statistics.stopTimer(statistics.typerNanos, start)
@@ -117,7 +123,10 @@ trait Analyzer extends AnyRef
           unit.body = typer.typed(unit.body)
           // interactive typed may finish by throwing a `TyperResult`
           if (!settings.Youtline.value) {
-            for (workItem <- unit.toCheck) workItem()
+            unit.toCheck foreach {
+              case now: CompilationUnit.ToCheckAfterUnit => now()
+              case later: CompilationUnit.ToCheckAfterTyper => toCheckAfterTyper += later
+            }
             if (!settings.isScaladoc && settings.warnUnusedImport)
               warnUnusedImports(unit)
             if (!settings.isScaladoc && settings.warnUnused.isSetByUser)

--- a/src/compiler/scala/tools/nsc/typechecker/Checkable.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/Checkable.scala
@@ -371,7 +371,7 @@ trait Checkable {
               val Xsym = X.typeSymbol
               val Psym = P.typeSymbol
               if (isSealedOrFinal(Xsym) && isSealedOrFinal(Psym) && (currentRun.compiles(Xsym) || currentRun.compiles(Psym)))
-                context.unit.toCheck += (() => recheckFruitless())
+                context.unit.addPostTyperCheck(() => recheckFruitless())
             }
         }
       }

--- a/src/compiler/scala/tools/nsc/typechecker/Namers.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/Namers.scala
@@ -1487,7 +1487,7 @@ trait Namers extends MethodSynthesis {
             context.warning(ddef.pos, msg, WarningCategory.OtherNullaryOverride)
             meth.updateAttachment(NullaryOverrideAdapted)
           }
-          context.unit.toCheck += (if (currentRun.isScala3) error _ else warn _)
+          context.unit.addPostUnitCheck(() => if (currentRun.isScala3) error() else warn())
           ListOfNil
         } else vparamSymss
       }

--- a/src/compiler/scala/tools/nsc/typechecker/Typers.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/Typers.scala
@@ -465,7 +465,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
         case _ =>
           QualifyingClassError(tree, qual)
           // Delay `setError` in namer, scala/bug#10748
-          if (immediate) setError(tree) else unit.toCheck += (() => setError(tree))
+          if (immediate) setError(tree) else unit.addPostUnitCheck(() => setError(tree))
           NoSymbol
       }
 
@@ -750,7 +750,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
         if (immediate) {
           action()
         } else {
-          unit.toCheck += (() => action(): Unit)
+          unit.addPostUnitCheck(() => action())
           true
         }
       }
@@ -3228,7 +3228,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
       namer.enterSyms(stats)
 
       // need to delay rest of typedRefinement to avoid cyclic reference errors
-      unit.toCheck += { () =>
+      unit.addPostUnitCheck(() => {
         val stats1 = typedStats(stats, NoSymbol)
         // this code kicks in only after typer, so `stats` will never be filled in time
         // as a result, most of compound type trees with non-empty stats will fail to reify
@@ -3238,7 +3238,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
         templ updateAttachment att.copy(stats = stats1)
         for (stat <- stats1 if stat.isDef && stat.symbol.isOverridingSymbol)
           stat.symbol setFlag OVERRIDE
-      }
+      })
     }
 
     def typedImport(imp : Import): Import = transformed.remove(imp) match {

--- a/test/junit/scala/tools/nsc/typechecker/TypedTreeTest.scala
+++ b/test/junit/scala/tools/nsc/typechecker/TypedTreeTest.scala
@@ -4,9 +4,39 @@ import org.junit.Assert.{assertEquals, assertNotEquals}
 import org.junit.Test
 
 import scala.tools.testkit.BytecodeTesting
+import scala.tools.testkit.BytecodeTesting._
 
 class TypedTreeTest extends BytecodeTesting {
   override def compilerArgs = "-Ystop-after:typer"
+
+  @Test def t12703(): Unit = {
+    import compiler._
+    val a =
+      """object Foo {
+        |  sealed trait A
+        |  sealed trait B
+        |  final case class C(x: Int) extends A with B
+        |}
+        |""".stripMargin
+    val b =
+      """object T {
+        |  import Foo._
+        |  def f(a: A) = a match {
+        |    case b: B => 0
+        |    case _ => 1
+        |  }
+        |}
+        |""".stripMargin
+
+    def check(a: String, b: String) = {
+      val r = newRun()
+      r.compileSources(makeSourceFile(a, "A.scala") :: makeSourceFile(b, "B.scala") :: Nil)
+      checkReport()
+    }
+
+    check(a, b)
+    check(b, a)
+  }
 
   @Test
   def keepBlockUndetparams(): Unit = {

